### PR TITLE
fix(static-file): pass producer as `Arc<Mutex<_>>` to ensure only one is active

### PR DESCRIPTION
**Pull request by shekhirin**
This PR fixes a critical bug with multiple `StaticFileProducer` instances writing to static files at the same time, potentially causing duplicate and inconsistent data.

We need to have at most one `StaticFileProducer` active at a time. Otherwise, such a situation is possible:
1. `StaticFileHook` is started in the engine loop via polling the next hook https://github.com/paradigmxyz/reth/blob/3d86c4ac593b342bfe81a3414fc27ff0d608a881/crates/consensus/beacon/src/engine/mod.rs#L1860-L1870
2. On the next engine loop iteration, sync controlled is polled https://github.com/paradigmxyz/reth/blob/3d86c4ac593b342bfe81a3414fc27ff0d608a881/crates/consensus/beacon/src/engine/mod.rs#L1839-L1840
3. Sync controller spawns the pipeline, and we call the `StaticFileProducer` before actually starting the execution of all stages https://github.com/paradigmxyz/reth/blob/3d86c4ac593b342bfe81a3414fc27ff0d608a881/crates/stages/src/pipeline/mod.rs#L185-L186
4. Both `StaticFileHook` and `Pipeline::produce_static_files` run the `StaticFileProducer`, and it writes to static files from both places.

---

Having only one `StaticFileProducer` can be ensured via sharing one copy between everyone using `Arc<Mutex<StaticFileProducer>>`, so that the user will need to lock the static file producer first for the duration of its usage. This PR introduces a newtype wrapper `StaticFileProducer(Arc<Mutex<StaticFileProducerInner>>)` that makes it easier to initialize and pass it around.

An additional measure that can be done to prevent the static files corruption is implemented https://github.com/paradigmxyz/reth/pull/7137.

### CODE REVIEW

File: crates/static-file/src/static_file_producer.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub type StaticFileProducerResult = RethResult<StaticFileTargets>;
 pub type StaticFileProducerWithResult<DB> = (StaticFileProducer<DB>, StaticFileProducerResult);
 
 /// Static File producer routine. See [StaticFileProducer::run] for more detailed description.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
```
Comment (mattsse) [inline crates/static-file/src/static_file_producer.rs:24]: good!

**Review by mattsse — COMMENTED** [review]
it's not super pretty, but this should do it

**Comment by shekhirin** [comment]
> it's not super pretty, but this should do it

can do it prettier via a type alias that's public and a private internal type

**Comment by mattsse** [comment]
> can do it prettier via a type alias

no aliases please :)

**Comment by shekhirin** [comment]
> > can do it prettier via a type alias
> 
> 
> 
> no aliases please :)

newtype

File: crates/consensus/beacon/src/engine/sync.rs [hunk 2]
```diff
@@ -468,13 +468,13 @@ mod tests {
 
             let provider_factory = create_test_provider_factory_with_chain_spec(chain_spec);
 
-            let static_file_producer = StaticFileProducer::new(
+            let snapshotter = StaticFileProducer::new(
```
Comment (joshieDo) [inline crates/consensus/beacon/src/engine/sync.rs:471]: `snapshotter` is a forbidden word now

Comment (shekhirin) [reply crates/consensus/beacon/src/engine/sync.rs:471]: HOW did it get here

Comment (shekhirin) [reply crates/consensus/beacon/src/engine/sync.rs:471]: (i use vim btw so probably that's how)

Comment (shekhirin) [reply crates/consensus/beacon/src/engine/sync.rs]: fixed

File: crates/static-file/Cargo.toml [hunk 3]
```diff
@@ -27,6 +27,7 @@ tokio-stream.workspace = true
 tracing.workspace = true
 clap = { workspace = true, features = ["derive"], optional = true }
 rayon.workspace = true
+parking_lot = { workspace = true, features = ["send_guard", "arc_lock"] }
```
Comment (joshieDo) [inline crates/static-file/Cargo.toml:30]: til

**Review by joshieDo — APPROVED** [review]

File: crates/static-file/Cargo.toml (hunk 3 above)
Comment (shekhirin) [reply crates/static-file/Cargo.toml:30]: same

**Review by mattsse — APPROVED** [review]

**Pull request merged**